### PR TITLE
Feature/type2 withdrawNST

### PR DESCRIPTION
**Pull request by bwhour**
## Summary by CodeRabbit

- **New Features**
  - Pectra-mode beacon withdrawals: partial & full requests, dynamic withdrawal-fee endpoint, pubkey-hash utility, excess-fee refunds, request/failure/refund events, strict validations, and controller forwarding to user capsules.

- **Tests**
  - Expanded Pectra-mode tests for partial/full flows, fee retrieval, pubkey hashing, validations, proofs, and edge cases.

- **Chores**
  - Added network precompile address constant, public minimum withdrawal fee, and placeholder bootstrap endpoints that currently revert as not yet supported.

### CODE REVIEW

File: src/core/NativeRestakingController.sol [hunk 1]
```diff
@@ -132,4 +132,38 @@ abstract contract NativeRestakingController is
...
+        capsule.requestFullWithdrawal{value: msg.value}(pubkey);
+    }
+
+    /// @notice Get current withdrawal fee for the caller's capsule
+    /// @return Current withdrawal fee in wei
+    function getCurrentWithdrawalFee() external view returns (uint256) {
+        IImuaCapsule capsule = _getCapsule(msg.sender);
+        return capsule.getCurrentWithdrawalFee();
+    }
+
```
[the remark this answers is not in the document]
Comment (bwhour) [reply src/core/NativeRestakingController.sol:168]: fixed.

File: src/core/Bootstrap.sol [hunk 2]
```diff
@@ -837,4 +837,19 @@
...
+    /// @notice Request full withdrawal from a validator via beacon chain (not supported in Bootstrap)
+    function requestBeaconFullWithdrawal(bytes calldata) external payable {
+        revert Errors.NotYetSupported();
+    }
+
+    /// @notice Get current withdrawal fee (not supported in Bootstrap)
+    function getCurrentWithdrawalFee() external pure returns (uint256) {
+        revert Errors.NotYetSupported();
+    }
+
```
[the remark this answers is not in the document]
Comment (bwhour) [reply src/core/Bootstrap.sol:854]: fixed.

File: src/core/ImuaCapsule.sol [hunk 3]
```diff
@@ -292,4 +332,134 @@ contract ImuaCapsule is ReentrancyGuardUpgradeable, ImuaCapsuleStorage, IImuaCap
...
+
+        // Encode according to EIP-7002 specification
+        // Per EIP-7002 example: abi.encodePacked(pubkey, amount) - Solidity handles endianness correctly
+        bytes memory callData = abi.encodePacked(
+            pubkey, // validator_pubkey (48 bytes)
+            uint64(amount) // amount (8 bytes) - Solidity encodes as expected by precompile
+        );
+
+        // Verify the input is exactly 56 bytes as required by EIP-7002 (48 bytes pubkey + 8 bytes amount)
+        require(callData.length == PectraConstants.CALLDATA_LENGTH, "ImuaCapsule: invalid calldata length");
```
[the remark this answers is not in the document]
Comment (bwhour) [reply src/core/ImuaCapsule.sol:430]: make sense. will remove it.

File: src/core/ImuaCapsule.sol [hunk 4]
```diff
@@ -292,4 +332,134 @@ contract ImuaCapsule is ReentrancyGuardUpgradeable, ImuaCapsuleStorage, IImuaCap
...
+
+        // Call the precompile with the fee provided by the caller
+        (bool success,) = PectraConstants.BEACON_WITHDRAWAL_PRECOMPILE.call{value: msg.value}(callData);
```
[the remark this answers is not in the document]
Comment (bwhour) [reply src/core/ImuaCapsule.sol:433]: will fix it.

File: src/libraries/PectraConstants.sol [hunk 5]
```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+/// @title PectraConstants
+/// @notice Constants for Pectra hard fork and EIP-7002 beacon withdrawal functionality
+/// @author imua-xyz
+library PectraConstants {
```
[the remark this answers is not in the document]
Comment (bwhour) [reply src/libraries/PectraConstants.sol:7]: Good point.  Just want to preserve the current storage layout without any modifications. Since these are Pectra-specific functions, placing them in a library keeps them independent and ensures no storage impact.

File: src/core/ImuaCapsule.sol [hunk 6]
```diff
@@ -292,4 +337,133 @@ contract ImuaCapsule is ReentrancyGuardUpgradeable, ImuaCapsuleStorage, IImuaCap
...
+            revert BeaconWithdrawalNotSupportedInPrePectraMode();
+        }
+        if (pubkey.length != PectraConstants.PUBKEY_LENGTH) {
+            revert InvalidValidatorPubkey(pubkey);
+        }
+        if (amount == 0) {
+            revert InvalidWithdrawalAmount(amount);
+        }
+
+        // Verify that the validator exists and is registered with this capsule
```
Comment (cloud8little) [inline src/core/ImuaCapsule.sol:360]: here validator pubkeyHash is not simple sha256 of pubkey, pubkeyHash is the first element in the validator container, it is generated by ssz rules.

**Comment by cloud8little** [comment]
Test failed: 

```
    ├─ [52058] 0x3ca48daA704F3f42B20c4DDbE5584e8c40Cf63D8::requestBeaconPartialWithdrawal{value: 1}(0x88e169e0a01cbcbfe2e5dc0abec6b504401a58ba34edeabd7f6939eb7c7cbb2730deb9da6ead98e260000c6582248545, 1000000000000000000 [1e18])
    │   ├─ [44894] 0x53BB4C117981A75BfC0aef3D7ae5408FBFb73d01::requestBeaconPartialWithdrawal{value: 1}(0x88e169e0a01cbcbfe2e5dc0abec6b504401a58ba34edeabd7f6939eb7c7cbb2730deb9da6ead98e260000c6582248545, 1000000000000000000 [1e18]) [delegatecall]
    │   │   ├─ [22828] 0x2c3c74f414f1CC81a2EeD3Ab3ac3B41C4C89097E::requestPartialWithdrawal{value: 1}(0x88e169e0a01cbcbfe2e5dc0abec6b504401a58ba34edeabd7f6939eb7c7cbb2730deb9da6ead98e260000c6582248545, 1000000000000000000 [1e18])
    │   │   │   ├─ [2308] 0x7be54abAA1b58b02bF5E084b8849F65d09B50Ee9::implementation() [staticcall]
    │   │   │   │   └─ ← [Return] 0xab05cfEF8C3b837409d5Fecb4b475bD022fEBa58
    │   │   │   ├─ [12710] 0xab05cfEF8C3b837409d5Fecb4b475bD022fEBa58::requestPartialWithdrawal{value: 1}(0x88e169e0a01cbcbfe2e5dc0abec6b504401a58ba34edeabd7f6939eb7c7cbb2730deb9da6ead98e260000c6582248545, 1000000000000000000 [1e18]) [delegatecall]
    │   │   │   │   ├─ [84] PRECOMPILES::sha256(0x88e169e0a01cbcbfe2e5dc0abec6b504401a58ba34edeabd7f6939eb7c7cbb2730deb9da6ead98e260000c6582248545) [staticcall]
    │   │   │   │   │   └─ ← [Return] 0xaa7391e5f302b78246e03d0f51f1799399484f8181b6580bc2880a5801d956fe
    │   │   │   │   └─ ← [Revert] UnregisteredValidator(0xaa7391e5f302b78246e03d0f51f1799399484f8181b6580bc2880a5801d956fe)
    │   │   │   └─ ← [Revert] UnregisteredValidator(0xaa7391e5f302b78246e03d0f51f1799399484f8181b6580bc2880a5801d956fe)
    │   │   └─ ← [Revert] UnregisteredValidator(0xaa7391e5f302b78246e03d0f51f1799399484f8181b6580bc2880a5801d956fe)
    │   └─ ← [Revert] UnregisteredValidator(0xaa7391e5f302b78246e03d0f51f1799399484f8181b6580bc2880a5801d956fe)
    └─ ← [Revert] UnregisteredValidator(0xaa7391e5f302b78246e03d0f51f1799399484f8181b6580bc2880a5801d956fe)
```

File: src/core/ImuaCapsule.sol [hunk 7]
```diff
@@ -292,4 +337,127 @@ contract ImuaCapsule is ReentrancyGuardUpgradeable, ImuaCapsuleStorage, IImuaCap
...
+        }
+
+        // Call the beacon withdrawal precompile
+        bool success = _callBeaconWithdrawalPrecompile(pubkey, amount);
+        if (!success) {
+            revert BeaconWithdrawalPrecompileFailed(pubkey, amount);
+        }
+
+        emit PartialWithdrawalRequested(pubkey, amount, capsuleOwner);
+    }
```
[the remark this answers is not in the document]
Comment (bwhour) [reply src/core/ImuaCapsule.sol:371]: @coderabbitai no need to change, all is the same.

File: src/core/ImuaCapsule.sol (hunk 6 above)
Comment (bwhour) [reply src/core/ImuaCapsule.sol:357]: yes. fixed.

File: src/core/ImuaCapsule.sol [hunk 8]
```diff
@@ -292,4 +337,132 @@ contract ImuaCapsule is ReentrancyGuardUpgradeable, ImuaCapsuleStorage, IImuaCap
...
+
+        // Verify that the validator exists and is registered with this capsule
+        bytes32 pubkeyHash = ValidatorContainer.computePubkeyHash(pubkey);
+        Validator storage validator = _capsuleValidators[pubkeyHash];
+        if (validator.status == VALIDATOR_STATUS.UNREGISTERED) {
+            revert UnregisteredValidator(pubkeyHash);
+        }
+
+        // Call the beacon withdrawal precompile
+        bool success = _callBeaconWithdrawalPrecompile(pubkey, amount);
```
Comment (cloud8little) [inline src/core/ImuaCapsule.sol:365]:
```suggestion
        bool success = _callBeaconWithdrawalPrecompile{value: msg.value}(pubkey, amount);
```
